core/src/main/java/org/apache/iceberg/avro/AvroWithPartnerSchemaVisitor.java

```diff
@@ -0,0 +1,188 @@
+      }
+    } else {
+      String[] fieldNames = visitor.structFieldNames(struct);
+      P[] fieldTypes = visitor.structFieldTypes(struct);
+      Preconditions.checkArgument(fieldTypes.length == fields.size(),
+          "Structs do not match: %s != %s", struct, record);
+      for (int i = 0; i < fieldTypes.length; i += 1) {
+        String fieldName = fieldNames[i];
+        Schema.Field field = fields.get(i);
+        Preconditions.checkArgument(AvroSchemaUtil.makeCompatibleName(fieldName).equals(field.name()),
+            "Structs do not match: field %s != %s", fieldName, field.name());
+        results.add(visit(fieldTypes[i], field.schema(), visitor));
+      }
+    }
+
+    visitor.recordLevels.pop();
+
+    return visitor.record(struct, record, names, results);
+  }
+
+  private static <P, T> T visitUnion(P type, Schema union, AvroWithPartnerSchemaVisitor<P, T> visitor) {
+    List<Schema> types = union.getTypes();
+    Preconditions.checkArgument(AvroSchemaUtil.isOptionSchema(union),
+        "Cannot visit non-option union: %s", union);
+    List<T> options = Lists.newArrayListWithExpectedSize(types.size());
+    for (Schema branch : types) {
+      if (branch.getType() == Schema.Type.NULL) {
+        options.add(visit(visitor.nullType(), branch, visitor));
+      } else {
+        options.add(visit(type, branch, visitor));
+      }
+    }
+    return visitor.union(type, union, options);
+  }
+
+  private static <P, T> T visitArray(P type, Schema array, AvroWithPartnerSchemaVisitor<P, T> visitor) {
+    if (array.getLogicalType() instanceof LogicalMap || visitor.isMapType(type)) {
+      Preconditions.checkState(
+          AvroSchemaUtil.isKeyValueSchema(array.getElementType()),
+          "Cannot visit invalid logical map type: %s", array);
+      List<Schema.Field> keyValueFields = array.getElementType().getFields();
+      return visitor.map(type, array,
+          visit(visitor.mapKeyType(type), keyValueFields.get(0).schema(), visitor),
+          visit(visitor.mapValueType(type), keyValueFields.get(1).schema(), visitor));
+
+    } else {
+      return visitor.array(type, array, visit(visitor.arrayElementType(type), array.getElementType(), visitor));
+    }
+  }
+
+  /**
+   * Just for checking state.
+   */
+  private Deque<String> recordLevels = Lists.newLinkedList();
+
+  // ---------------------------------- Partner type methods ---------------------------------------------
+
+  public boolean schemaEvolution() {
+    return false;
+  }
+
+  public abstract boolean isMapType(P type);
+
+  public abstract boolean isValidMapKey(P type);
+
+  public abstract P arrayElementType(P arrayType);
+
+  public abstract P mapKeyType(P mapType);
+  public abstract P mapValueType(P mapType);
+
+  public String[] structFieldNames(P structType) {
+    throw new UnsupportedOperationException();
+  }
+
+  public P[] structFieldTypes(P structType) {
```

Comment:
Arrays are a bit difficult to work with. For the structure-based traversal, how about combining `structFieldNames` and `structFieldTypes` into a single method indexed by position in the struct?

```java
  Pair<String, P> fieldNameAndType(P structType, int pos);
```

That corresponds to the `fieldType(int id)` for the id-based lookup.